PR to merge Develop branch into Master

SebRenon opened the pull request:
First version of the code to parse a website and return all its email addresses

mmcmahon-jana commented on src/main/java/com/sebrenon/emailcrawler/util/CrawlerUtils.java:
```diff
@@ -0,0 +1,88 @@
+
+    private final static String PROTOCOL_HTTP = "http://";
+    private final static String PROTOCOL_HTTPS = "https://";
+    private final static String MALFORMED_PROTOCOL = "http:////";
```
Was this from a specific site you tested with, or just a general pattern you have encountered before.

SebRenon replied:
It's actually from http://www.jana.com/home, the link surrounding the logo, on the top left of the page.

`
<a href="http:////www.jana.com/home" id="hs-link-module_148274919024855" style="border-width:0px;border:0px;"><img src="https://www.jana.com/hubfs/Logos/logo.svg?t=1506710710712" class="hs-image-widget " style="width:140px;border-width:0px;border:0px;" width="140" alt="JANA" title="JANA"></a>
`

mmcmahon-jana replied:
Ha nice, thanks.  We'll have to get someone to look into that.

mmcmahon-jana commented on src/main/java/com/sebrenon/emailcrawler/Main.java:
```diff
@@ -0,0 +1,117 @@
+            }
+            bufferedReader.close();
+        } catch (Exception e) {
+            System.out.println(String.format("\nError for url %s with msg: %s", urlString, e.getMessage()));
```
When crawling ` https://www.washingtonpost.com`, I see a lot of errors like this
`
Error for url https://www.washingtonpost.com/world/europe/macedonia-blocks-migrants-on-border-with-greece/2015/08/20/fa2b2b7e-4734-11e5-9f53-d1e3ddfd0cda_story.html with msg: https://www.washingtonpost.com/world/europe/macedonia-blocks-migrants-on-border-with-greece/2015/08/20/fa2b2b7e-4734-11e5-9f53-d1e3ddfd0cda_story.html
`
which seem to correspond with 404's, which seem safe to ignore.  Can you handle those and other HTTP errors in a way that doesn't generate the output?

SebRenon replied:
Ok, I pushed a change: 6d2299d64aa210ae3bafb6d286fd0fd4cac0e0b9

mmcmahon-jana commented:
I know we did not specify testing with a site like `https://www.washingtonpost.com`, but running your crawler seems to run for a an extended amount of time (I stopped it after a while).  What methods could you use to try to limit the crawler so that it doesn't attempt to crawl the entire site.
I'm thinking of things like limiting the depth you crawl, or even the overall number of pages.

SebRenon commented:
@mmcmahon-jana I pushed the code to handle Javascript emails: 3d7d39477245ce4483d3a0c95cfb63201a00e048

I updated the regex to look for '@' or '[at]' and then I added some logic to replace the '[at]' by ’@'. Let me know what you think.

Thanks

SebRenon commented:
@mmcmahon-jana To limit the crawling, I imagine I could limit it to a specific level _n_, meaning I would only authorize crawling up to _n_ pages from the starting domain:
Root = Lvl 0
Children(Links found in Root) = Lvl 1
GrandChildren (Links found in Links from Root) = Lvl 2
...
Last level = Lvl _n_

I could also limit the amount of url to crawl but that won't return the same data every time, depending on the page's time of response...
Same thing if I set a time limit for the whole process (e.g: stop after 30sec). The final result won't be consistent.

Do you want me to implement the depth limitation?

mmcmahon-jana commented:
Yeah, i I'd like to see your implementation of limiting the crawl to prevent that kind of never ending search. 
Thanks!

SebRenon commented:
@mmcmahon-jana I added the crawl limitation: 6dafe28c520c3f530cd446d660da27665904486a
